File: daemon/internal/connectors/slack/smoke_test.go
package slack

import (
	"testing"
	"time"
)

func TestBuildSmokeEvidenceStructuredSkipAndFakePass(t *testing.T) {
	t.Parallel()

	now := time.Date(2026, 5, 8, 13, 0, 0, 0, time.UTC)
	skip := BuildSmokeEvidence(SmokeInput{
		TenantID:           "ten_slack",
		ConnectorID:        "slack-main",
		WorkspaceBindingID: "workspace_binding_redacted",
		ValidatedAt:        now,
	})
	if skip.Status != SmokeSkipped || skip.AuthorizationMode != AuthorizationModeUnavailable || skip.RemainingRisk == "" {
		t.Fatalf("unexpected structured skip: %+v", skip)
	}

	fakePass := BuildSmokeEvidence(SmokeInput{
		TenantID:           "ten_slack",
		ConnectorID:        "slack-main",
		WorkspaceBindingID: "workspace_binding_redacted",
		FakeOAuth:          true,
		Passed:             true,
		ValidatedAt:        now,
		SafeEvidence:       map[string]string{"transport": "fake"},
		RemainingRisk:      "live provider not exercised",
	})
	if fakePass.Status != SmokePassed || fakePass.AuthorizationMode != AuthorizationModeFakeOAuth || fakePass.Reason != "healthy" {
		t.Fatalf("unexpected fake pass: %+v", fakePass)
	}
}

func TestBuildSmokeEvidenceSuppressesUnsafeEvidence(t *testing.T) {
	t.Parallel()

	smoke := BuildSmokeEvidence(SmokeInput{
		TenantID:           "ten_slack",
		ConnectorID:        "slack-main",
		WorkspaceBindingID: "workspace_binding_redacted",
		SafeLiveApproved:   true,
		SafeEvidence: map[string]string{
			"authorization": "Bearer xoxb-secret",
		},
	})
	if smoke.RedactionStatus != "suppressed" || len(smoke.SafeEvidence) != 0 {
		t.Fatalf("expected unsafe smoke evidence to be suppressed, got %+v", smoke)
	}
}

File: crates/channels/slack/tests/smoke_test.rs
//! Behavioral tests for the smoke evidence builder (port of smoke_test.go).

use std::collections::HashMap;

use chrono::{TimeZone, Utc};
use kura_slack::smoke::{
    AuthorizationMode, SmokeEvidence, SmokeInput, SmokeStatus, build_smoke_evidence,
};

fn now() -> chrono::DateTime<Utc> {
    Utc.with_ymd_and_hms(2026, 5, 8, 13, 0, 0)
        .single()
        .expect("valid timestamp")
}

#[test]
fn build_smoke_evidence_structured_skip_and_fake_pass() {
    let skip = build_smoke_evidence(SmokeInput {
        tenant_id: "ten_slack".to_string(),
        connector_id: "slack-main".to_string(),
        workspace_binding_id: "workspace_binding_redacted".to_string(),
        validated_at: now(),
        ..SmokeInput::default()
    });
    assert_eq!(skip.status, SmokeStatus::Skipped);
    assert_eq!(skip.authorization_mode, AuthorizationMode::Unavailable);
    assert!(
        !skip.remaining_risk.is_empty(),
        "structured skip must carry remaining risk"
    );

    let fake_pass = build_smoke_evidence(SmokeInput {
        tenant_id: "ten_slack".to_string(),
        connector_id: "slack-main".to_string(),
        workspace_binding_id: "workspace_binding_redacted".to_string(),
        fake_oauth: true,
        passed: true,
        validated_at: now(),
        safe_evidence: HashMap::from([("transport".to_string(), "fake".to_string())]),
        remaining_risk: "live provider not exercised".to_string(),
        ..SmokeInput::default()
    });
    assert_eq!(fake_pass.status, SmokeStatus::Passed);
    assert_eq!(fake_pass.authorization_mode, AuthorizationMode::FakeOAuth);
    assert_eq!(fake_pass.reason, "healthy");
}

#[test]
fn build_smoke_evidence_suppresses_unsafe_evidence() {
    let smoke = build_smoke_evidence(SmokeInput {
        tenant_id: "ten_slack".to_string(),
        connector_id: "slack-main".to_string(),
        workspace_binding_id: "workspace_binding_redacted".to_string(),
        safe_live_approved: true,
        safe_evidence: HashMap::from([(
            "authorization".to_string(),
            "Bearer xoxb-secret".to_string(),
        )]),
        ..SmokeInput::default()
    });
    assert_eq!(smoke.redaction_status.as_str(), "suppressed");
    assert!(
        smoke.safe_evidence.is_empty(),
        "unsafe smoke evidence must be suppressed"
    );
    let _: SmokeEvidence = smoke;
}
